src/goto-harness/memory_snapshot_harness_generator.cpp

```diff
@@ -69,6 +76,66 @@ void memory_snapshot_harness_generatort::handle_option(
   }
 }
 
+optionalt<goto_programt::const_targett>
+memory_snapshot_harness_generatort::get_start_instruction(
+  const goto_modelt &goto_model)
+{
+  const auto &goto_functions = goto_model.goto_functions;
+  if(line_number.has_value())
+  {
+    // by line: iterate over all instructions until source location match
+    for(const auto &entry : goto_functions.function_map)
+    {
+      const auto &goto_function = entry.second;
+
+      if(!goto_function.body_available())
+        continue;
+      const auto &goto_program = goto_function.body;
+
+      const auto it = std::find_if(
+        goto_program.instructions.begin(),
+        goto_program.instructions.end(),
+        [this](const goto_programt::instructiont &instruction) {
+          return safe_string2unsigned(id2string(
+                   instruction.source_location.get_line())) >= *line_number;
+        });
+
+      if(it != goto_program.instructions.end())
+      {
+        entry_function_name = entry.first; //TODO - move elsewhere
+        return it;
+      }
+    }
+  }
+  else if(!entry_function_name.empty())
+  {
+    // by function(+location): search for the function then jump to n-th
+    // location, then check the number
+    const auto &goto_function =
+      goto_functions.function_map.find(entry_function_name);
+    if(
+      goto_function != goto_functions.function_map.end() &&
+      goto_function->second.body_available())
+    {
+      const auto &goto_program = goto_function->second.body;
+      if(!location_number.has_value())
+        return goto_program.instructions.begin();
+      if(goto_program.instructions.size() > *location_number)
```

Comment:
Because I'm advancing the iterator by `*location_number` position and I don't want to get past the `end()`.

Reply:
Ah, sorry, I missed that. But also this isn't how the `location_number` member of instructions works: these location numbers are guaranteed to be unique across _all_ goto functions, and monotonically increasing in a single function. Therefore a location number could well be larger than the size of a single function. At the same time, there could be gaps, so locations in a function could legitimately be numbered "7, 10, 15" instead of "1, 2, 3."

Reply:
I give up. What about `find_if` to get the right instruction? That should be safe right?

Reply:
Yes, I'm afraid you will have to do some form of search (technically binary search would be faster, but quite likely no one will notice the cost of linear search here).

Reply:
But would binary search on a `std::list` really be faster? I thought that random access there is linear in the `distance` of the jump.

Reply:
That's a question of how expensive the comparison operation is. Most likely you are right in that binary search would not add any value (or may indeed be more expensive).